src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertForEachToLinqQueryProvider)), Shared]
+    internal sealed class CSharpConvertForEachToLinqQueryProvider : AbstractConvertForEachToLinqQueryProvider<ForEachStatementSyntax, StatementSyntax>
+    {
+        protected override ForEachStatementSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+            => root.FindNode(span) as ForEachStatementSyntax;
+
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override string Title => CSharpFeaturesResources.Convert_to_query;
+
+        protected override IConverter CreateDefaultConverter(
+            ForEachStatementSyntax forEachStatement,
+            ImmutableArray<SyntaxNode> convertingNodes,
+            ImmutableArray<SyntaxToken> identifiers,
+            ImmutableArray<StatementSyntax> statements)
+            => new DefaultConverter(forEachStatement, convertingNodes, identifiers, statements);
+
+        // Do not try to refactor queries with comments or conditional compilation in them.
+        // We can consider supporting queries with comments in the future.
+        protected override bool Validate(StatementSyntax statement)
+         => !(statement.ContainsDirectives ||
+            statement.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+             SyntaxKind.SingleLineCommentTrivia,
+             SyntaxKind.MultiLineCommentTrivia,
+             SyntaxKind.MultiLineDocumentationCommentTrivia)));
```

Comment:
i still think we sholud not have this check.  absolutely no other IDE refactorings do the comment trivia checks.

Note: i'm ok leaving in the directives check (since handling those might be difficult).

Reply:
Working on this now. If we just remove the condition without any extra code, we keep comments assigned to tokens or nodes we preserve but lose comments from nodes we are removing.

For example, we remove "from", "(" and ")" from "from (var x in a)". Therefore, all comments just after ")" will be lost. Now I'm considering if we can properly save and rearrange comments.

---
In reply to: [191058187](https://github.com/dotnet/roslyn/pull/27130#discussion_r191058187) [](ancestors = 191058187)